lib/travis/logs/services/timing_info.rb

```diff
@@ -0,0 +1,132 @@
+# frozen_string_literal: true
+
+require 'concurrent'
+require 'date'
+require 'travis/logs'
+
+module Travis
+  module Logs
+    module Services
+      class TimingInfo
+        attr_reader :job_id, :database
+
+        private :database
+
+        TIMER_START = /travis_time:start:(?<timer_id>[0-9a-f]+)/
+        TIMER_END = /travis_time:end:(?<timer_id>[0-9a-f]+):(?<info>[^\r]+)\r/
+
+        def self.run
+          new.run
+        end
+
+        def self.timing_info(job_id)
+          new.timing_info(job_id)
+        end
+
+        def initialize(job_id, database: Travis::Logs.database_connection)
+          @job_id   = job_id
+          @database = database
+        end
+
+        def run
+          timing_info job_id
+        end
+
+        def timing_info(job_id)
+          timer_stack = []
+          # Build a honeycomb event
+          honey = Travis::Honeycomb.honey
+
+          ev_builder = honey.builder
+          ev_builder.writekey = Travis.config.logs.honeycomb.build_timings.writekey
+          ev_builder.dataset  = Travis.config.logs.honeycomb.build_timings.dataset
+
+          content.each_line do |l|
+            l.scan(/#{TIMER_START}|#{TIMER_END}/) do |start_timer_id, end_timer_id, info|
+              if start_timer_id
+                timer_stack << start_timer_id
+                next
+              end
+
+              if timer_stack.empty?
+                Travis.logger.warn "Empty log event timer stack"
+                next
+              end
+              unless (last_timer_id = timer_stack.pop) == end_timer_id
+                Travis.logger.warn "Timer IDs do not match. From stack: #{last_timer_id}; seen: #{end_timer_id}"
+                next
+              end
+
+              # matched TIMER_END regexp, so we have `end_timer_id` and `info` defined
+              marker_data = parse_marker_data(info)
+              next unless extra_info?(marker_data)
+
+              event = ev_builder.event
+              event.add_field(:job_id, job_id)
+              event.add normalize_timestamps(marker_data)
+
+              event.send
+              Travis.logger.debug event.to_s
+            end
+          end
+        end
+
+        def log
+          @log ||= begin
+            log_id = database.log_id_for_job_id(job_id)
+            log = database.log_for_id(log_id)
+            unless log
+              Travis.logger.warn(
+                'log not found',
+                action: 'timing_info', id: log_id, result: 'not_found'
+              )
+            end
+            log
+          end
+        end
+        alias fetch log
```

Comment:
i don't see the alias `fetch` used? if it isn't i'd personally split out fetching the log to the method `fetch`, and then memoize it in `log` using just one line `@log ||= fetch` ... mostly because i dislike `begin/end` blocks. i'd also split out `log_not_found` and get rid of the `unless/end` block using `log_not_found unless log` ... just a different style though, obviously

Reply:
We can remove `alias`, for sure. This method was originally a copy-paste from https://github.com/travis-ci/travis-logs/blob/849119aec39c21c06c14086301d9f67b26a7930f/lib/travis/logs/services/archive_log.rb#L61-L74, where the repeated use of the name `log` can be quite confusing.